warden/x/async/keeper/keeper.go

```diff
@@ -289,3 +289,48 @@ func (k Keeper) callEVMWithData(
 
 	return res, nil
 }
+
+type completedFuture struct {
+	future types.Future
+	result types.FutureResult
+}
+
+func (k Keeper) getCompletedFuturesWithoutValidatorVote(ctx context.Context, valAddress []byte, limit int) ([]completedFuture, error) {
+	it, err := k.futures.results.IterateRaw(ctx, nil, nil, collections.OrderAscending)
+	if err != nil {
+		return nil, err
+	}
+	defer it.Close()
+
+	futures := make([]completedFuture, 0, limit)
+	for ; it.Valid(); it.Next() {
+		id, err := it.Key()
+		if err != nil {
+			return nil, err
+		}
+
+		result, err := it.Value()
+		if err != nil {
+			return nil, err
+		}
+
+		if _, err := k.votes.Get(ctx, collections.Join(id, valAddress)); err == nil {
```

Comment:
we can save some work by just checking if the key exists instead of loading/decoding its value just to discard it

```suggestion
		if found, err := k.votes.Has(ctx, collections.Join(id, valAddress)); found {
```

Reply:
+